README.md
# Lazyload Responsive

If you're lazyloading images, you might as well add features to load responsive images while you're at it

## Getting Started
Download the [production version][min] or the [development version][max].

[min]: https://raw.github.com/kevinkirchner/lazyload-responsive/master/dist/lazyload-responsive.min.js
[max]: https://raw.github.com/kevinkirchner/lazyload-responsive/master/dist/lazyload-responsive.js

In your web page:

```html
<script src="jquery.js"></script>
<script src="dist/lazyload-responsive.min.js"></script>
<script>
jQuery(function($) {
  $.awesome(); // "awesome"
});
</script>
```

## Documentation
_(Coming soon)_

## Examples
_(Coming soon)_

## Release History

### 0.2.2

- Putting a leash on requests
- Load images on scroll
- Load images on resize
- Added grunt; run `npm install` before sending me a pull request


### 0.2 - Ragged Dog

- Working, but making http requests like we're living in the 90's


### 0.1 - Unashamed Fox

- Working, but making http requests like we're living in 2050 A.D.

CHANGELOG
# Change Log


## 0.2.2

- Putting a leash on requests
- Load images on scroll
- Added grunt; run `npm install` before sending me a pull request


## 0.2 - Ragged Dog

- Working, but making http requests like we're living in the 90's


## 0.1 - Unashamed Fox

- Working, but making http requests like we're living in 2050 A.D.

test/lazyload-responsive_test.js
(function($) {
  /*
    ======== A Handy Little QUnit Reference ========
    http://api.qunitjs.com/

    Test methods:
      module(name, {[setup][ ,teardown]})
      test(name, callback)
      expect(numberOfAssertions)
      stop(increment)
      start(decrement)
    Test assertions:
      ok(value, [message])
      equal(actual, expected, [message])
      notEqual(actual, expected, [message])
      deepEqual(actual, expected, [message])
      notDeepEqual(actual, expected, [message])
      strictEqual(actual, expected, [message])
      notStrictEqual(actual, expected, [message])
      throws(block, [expected], [message])
  */

    // Lazyload-Responsive
    module('LazyloadResponsive', {
        setup: function() {
            this.elems = $('img[data-lzld-src]');
        }
    });
    
    test('is chainable', function() {
      expect(1);
      // Not a bad test to run on collection methods.
      strictEqual(this.elems.lazyloadResponsive(), this.elems, 'should be chainable');
    });
    
    module('Request Management', {
        setup: function() {
            var that = this;
            that.requestedSrcs = [];
            // @see http://stackoverflow.com/questions/3596583/javascript-detect-an-ajax-event
            that.ajaxListener = new Object();
            that.ajaxListener.tempOpen = XMLHttpRequest.prototype.open;
            that.ajaxListener.tempSend = XMLHttpRequest.prototype.send;
            that.ajaxListener.callback = function () {
                // TODO: if method was one to look for an image, then run this
                if (true) {
                    that.requestedSrcs.push(this.url);
                }
              // this.method :the ajax method used
              // this.url    :the url of the requested script (including query string, if any) (urlencoded) 
              // this.data   :the data sent, if any ex: foo=bar&a=b (urlencoded)
            }

            XMLHttpRequest.prototype.open = function(a,b) {
              if (!a) var a='';
              if (!b) var b='';
              that.ajaxListener.tempOpen.apply(this, arguments);
              that.ajaxListener.method = a;  
              that.ajaxListener.url = b;
              if (a.toLowerCase() == 'get') {
                that.ajaxListener.data = b.split('?');
                that.ajaxListener.data = that.ajaxListener.data[1];
              }
            }

            XMLHttpRequest.prototype.send = function(a,b) {
              if (!a) var a='';
              if (!b) var b='';
              that.ajaxListener.tempSend.apply(this, arguments);
              if(that.ajaxListener.method.toLowerCase() == 'post')that.ajaxListener.data = a;
              that.ajaxListener.callback();
            }
        }
    });

    // No file should ever be requested more than once
    asyncTest('has no duplicate image requests', 1, function(){
        // TODO: make this an on method that makes sense
        that.ajaxListener.on('readystatechange', function(data){
            QUnit.log(data)
            start();
        })
    })

// Event: Onload
// Images within range should be loaded AND resized

// Event: Scroll
// Images within range should be loaded AND resized
// Call on scroll should be throttled
// Option related tests

// Event: Resize
// Images within range should be loaded AND resized
// Call on resize should be throttled
// Option related tests

// Dynamically Added Content
// Images within range should be loaded AND resized


/*
  module('jQuery#awesome', {
    // This will run before each test in this module.
    setup: function() {
      this.elems = $('#qunit-fixture').children();
    }
  });

  test('is chainable', function() {
    expect(1);
    // Not a bad test to run on collection methods.
    strictEqual(this.elems.lazyloadResponsive(), this.elems, 'should be chainable');
  });

  test('is awesome', function() {
    expect(1);
    strictEqual(this.elems.lazyloadResponsive().text(), 'awesome0awesome1awesome2', 'should be awesome');
  });

  module('jQuery.lazyloadResponsive');

  test('is awesome', function() {
    expect(2);
    strictEqual($.lazyloadResponsive(), 'awesome.', 'should be awesome');
    strictEqual($.lazyloadResponsive({punctuation: '!'}), 'awesome!', 'should be thoroughly awesome');
  });

  module(':awesome selector', {
    // This will run before each test in this module.
    setup: function() {
      this.elems = $('#qunit-fixture').children();
    }
  });

  test('is awesome', function() {
    expect(1);
    // Use deepEqual & .get() when comparing jQuery objects.
    deepEqual(this.elems.filter(':awesome').get(), this.elems.last().get(), 'knows awesome when it sees it');
  });
*/

}(jQuery));


Gruntfile.js
'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON('lazyload-responsive.jquery.json'),
    banner: '/*! <%= pkg.title || pkg.name %> - v<%= pkg.version %> - ' +
      '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
      '<%= pkg.homepage ? "* " + pkg.homepage + "\\n" : "" %>' +
      '* Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
      ' Licensed <%= _.pluck(pkg.licenses, "type").join(", ") %> */\n',
    // Task configuration.
    clean: {
      files: ['dist']
    },
    concat: {
      options: {
        banner: '<%= banner %>',
        stripBanners: true
      },
      dist: {
        src: ['src/<%= pkg.name %>.js'],
        dest: 'dist/<%= pkg.name %>.js'
      },
    },
    uglify: {
      options: {
        banner: '<%= banner %>'
      },
      dist: {
        src: '<%= concat.dist.dest %>',
        dest: 'dist/<%= pkg.name %>.min.js'
      },
    },
    qunit: {
      files: ['test/**/*.html']
    },
    jshint: {
      gruntfile: {
        options: {
          jshintrc: '.jshintrc'
        },
        src: 'Gruntfile.js'
      },
      src: {
        options: {
          jshintrc: 'src/.jshintrc'
        },
        src: ['src/**/*.js']
      },
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/**/*.js']
      },
    },
    watch: {
      gruntfile: {
        files: '<%= jshint.gruntfile.src %>',
        tasks: ['jshint:gruntfile']
      },
      src: {
        files: '<%= jshint.src.src %>',
        tasks: ['jshint:src', 'qunit']
      },
      test: {
        files: '<%= jshint.test.src %>',
        tasks: ['jshint:test', 'qunit']
      },
    },
  });

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-qunit');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');

  // Default task.
  grunt.registerTask('default', ['jshint', 'qunit', 'clean', 'concat', 'uglify']);

};


src/contentLoaded.js
/*!
* contentloaded.js
*
* Author: Diego Perini (diego.perini at gmail.com)
* Summary: cross-browser wrapper for DOMContentLoaded
* Updated: 20101020
* License: MIT
* Version: 1.2
*
* URL:
* http://javascript.nwbox.com/ContentLoaded/
* http://javascript.nwbox.com/ContentLoaded/MIT-LICENSE
*
*/

// @win window reference
// @fn function reference
function contentLoaded(win, fn) {

    var done = false, top = true,

    doc = win.document, root = doc.documentElement,

    add = doc.addEventListener ? 'addEventListener' : 'attachEvent',
    rem = doc.addEventListener ? 'removeEventListener' : 'detachEvent',
    pre = doc.addEventListener ? '' : 'on',

    init = function(e) {
        if (e.type == 'readystatechange' && doc.readyState != 'complete') return;
        (e.type == 'load' ? win : doc)[rem](pre + e.type, init, false);
        if (!done && (done = true)) fn.call(win, e.type || e);
    },

    poll = function() {
        try { root.doScroll('left'); } catch(e) { setTimeout(poll, 50); return; }
        init('poll');
    };

    if (doc.readyState == 'complete') fn.call(win, 'lazy');
    else {
        if (doc.createEventObject && root.doScroll) {
            try { top = !win.frameElement; } catch(e) { }
            if (top) poll();
        }
        doc[add](pre + 'DOMContentLoaded', init, false);
        doc[add](pre + 'readystatechange', init, false);
        win[add](pre + 'load', init, false);
    }

}

src/lazyload-responsive.raw.js
/*!
* lazyload-responsive.js
*
* Author: Kevin Kirchner (kirchner.kevin at gmail.com)
* Summary: cross-browser lazy and responsive image loader
* Version: 0.2.1
*
* URL:
* https://github.com/kevinkirchner/lazyload-responsive
*
*/

(function(window, document){
    
    // Set default configuration
    var config = {
        srcAttr: 'data-lzld-src',
        offset: 200, // distance (px) below the fold where images will be loaded
        highResThreshold: 1.5, // any pixel ratio >= this number will be flagged as high-res
        throttleInterval: 20, // throttled interval for scroll and resize events
        useGlobalImgConfig: false, // if `false`, the script will look for the following imgConfig on each lzld img (e.g. img.getAttribute('data-lzld-highressuffix') || imgConfig.highressuffix - which takes the script longer to process); setting to `true` is the fastest option
        findSmallerImgs: true, // forces script to look for and load smaller images as viewport is resized to a smaller size
        imgConfig: {
            highressuffix: '@2x', // e.g. imagename@2x.jpg or imagename_400@2x.jpg would be the high-res images
            loadevent: ['scroll','resize'], // you may want to load the images on a custom event; this is where you do that
            longfallback: true, // will look for all smaller images before loading the original (and largest image)
            lowres: false, // forces script to **not** look for high-res images
            sizedown: false, // by default images are sized up; this option forces it to get an image larger than the viewport and shrink it; NOTE: setting to `true` will load larger images and increase pageload
            widthfactor: 200 // looks for images with the following naming convention [real-image-src]_[factor of widthfactor].[file-extenstion]
        }
    };
    
    var domReady = false;
    contentLoaded(window,function(){ domReady = true; });
    
    var LazyloadResponsive = {
        // Options
        _o: config,
        // Flags
        _f: {},
        // Init variables
        imgs: [],
        loadImgsQ: [], // All images start here
        loadedImgs: [], // All loaded (and correctly sized) images end here; Also doubles as the array of images that need to be checked when browser is resized
        requestingBaseImgSrc: [], // Store base file name (e.g. 'path/to/image.jpg') of image srcs currently being requested to prevent multiple requests for different sizes of the same image
        availableImgSrc: [], // Store all successful requests to reuse when needed
        unavailableImgSrc: [], // Store all unsuccessful requests to prevent an unnecessary 404 request
        xmlhttp: [],
        resizeTimeout: 0,
        initialize: function() {
            var that = this;
            // Set high-res flag
            that._f.isHighRes = that._u.getPixelRatio() >= that._o.highResThreshold;
            // Collect images
            that.collectImgs();
            // Load first batch of images
            that.loadImgs();
            // Attach Events
            that.attachEvents();
        },
        // This can be run after adding new images to the page to make them lazyload-responsive
        collectImgs: function() {
            var that = this;
            var allImgs = document.getElementsByTagName( "img" );
            for ( var i = 0, il = allImgs.length; i < il; i++ ){
                if ( allImgs[i].getAttribute( that._o.srcAttr ) !== null ) {
                    var img = allImgs[i];
                    if (that._u.inArray(img, that.imgs) === -1) {
                        that.imgs.push( img );
                    }
                    if (!img.getAttribute('data-lzld-complete') && that._u.isVisible(img)) {
                        that.loadImgsQ.push( img );
                    }
                }
            }
        },
        collectLoadImgsQ: function() {
            var that = this;
            for ( var i = 0, il = that.imgs.length; i < il; i++ ){
                var img = that.imgs[i];
                if ( that._u.inArray(img, that.loadImgsQ) === -1 && !img.getAttribute('data-lzld-complete') && that._u.isVisible(img) ) {
                    that.loadImgsQ.push( img );
                }
            }
        },
        loadImgs: function() {
            // start with loadImgsQ
            var that = this;
            // reverse the img queue
            that.loadImgsQ.reverse();
            for (var i = that.loadImgsQ.length - 1; i >= 0; i--) {
                var img = that.loadImgsQ[i];
                var imgData = that.getImgData(img);
                // begin the request(s)
                that.requestImg(imgData);
                // remove from loadImgsQ
                that.loadImgsQ.splice(i,1);
            }
        },
        resizeImgs: function() {
            var that = this;
            for (var i = 0, il = that.loadedImgs.length; i < il; i++){
                var imgData = that.loadedImgs[i];
                // reset the alias array b/c of the new browser width
                imgData.aliasArray = that.getAliasArray(imgData._e.img, imgData._o);
                
                // only request an image when alias array changes
                // be sure to stop other requests being made currently before requesting again
                that.requestImg(imgData);
            }
        },
        attachEvents: function() {
            var that = this;
            var winW = that._u.getViewport("Width");
            var throttleLoad = that._u.throttle( function(){
                that.collectLoadImgsQ();
                that.loadImgs();
            }, that._o.throttleInterval);
            var throttleResize = that._u.throttle( function(){
                that.resizeImgs();
            }, that._o.throttleInterval);

            // onresize
            that._u.addEvent(window,"resize",function(){
                var newWinW = that._u.getViewport("Width");
                if (that._o.findSmallerImgs || newWinW > winW) {
                    throttleResize();
                }
                // TODO: when the browser is resized and images change, new images MAY come into view that need to be loaded, we should check for those after the resizing is complete
                winW = newWinW;
            });
            // onscroll
            that._u.addEvent(window,"scroll",function(){
                throttleLoad();
            });
        },
        // Should only be run once per image, otherwise it needs to be stored. imgData.aliasArray is the only one that should need to change
        getImgData: function(img) {
            var that = this;
            var imgConfig = that.getImgConfig(img);
            var parseImgSrc = img.getAttribute('data-lzld-isparsed') ? that.getStoredParsedImgSrc(img) : that.getParsedImgSrc(img);
            var aliasArray = that.getAliasArray(img, imgConfig);
            var imgData = {
                // store the element
                _e: { img: img },
                // image config options
                _o: imgConfig,
                // get parsed img src
                _p: parseImgSrc,
                // calculate array of names to try with the current browser width and image options taken into account
                aliasArray: aliasArray
            };
            return imgData;
        },
        getImgConfig: function(img) {
            var that = this;
            if (that._o.useGlobalImgConfig) { return that._o.imgConfig; }
            // else look for options stored on the <img>
            return {
                highressuffix:  img.getAttribute('data-lzld-highressuffix') || that._o.imgConfig.highressuffix,
                loadevent:      img.getAttribute('data-lzld-loadevent') || that._o.imgConfig.loadevent,
                longfallback:   img.getAttribute('data-lzld-longfallback') || that._o.imgConfig.longfallback,
                lowres:         img.getAttribute('data-lzld-lowres') || that._o.imgConfig.lowres,
                sizedown:       img.getAttribute('data-lzld-sizedown') || that._o.imgConfig.sizedown,
                widthfactor:    img.getAttribute('data-lzld-widthfactor') || that._o.imgConfig.widthfactor
            };
        },
        getStoredParsedImgSrc: function(img) {
            return {
                filePath: img.getAttribute('data-lzld-filepath'),
                fileName: img.getAttribute('data-lzld-filename'),
                fileExt: img.getAttribute('data-lzld-fileext'),
                aliasBase: img.getAttribute('data-lzld-aliasbase'),
                baseSrc: img.getAttribute('data-lzld-basesrc')
            };
        },
        getParsedImgSrc: function(img) {
            var that = this,
                imgSrc = img.getAttribute( that._o.srcAttr ),
                lastSlash = imgSrc.lastIndexOf('/')+1,
                filePath = imgSrc.substring(0,lastSlash),
                file = imgSrc.substring(lastSlash),
                lastPeriod = file.lastIndexOf('.'),
                fileName = file.substring(0,lastPeriod),
                fileExt = file.substring(lastPeriod);
            // Store on the image for later use if needed
            img.setAttribute('data-lzld-filepath', filePath);
            img.setAttribute('data-lzld-filename', fileName);
            img.setAttribute('data-lzld-fileext', fileExt);
            img.setAttribute('data-lzld-aliasbase', filePath + fileName);
            img.setAttribute('data-lzld-basesrc', filePath + fileName + fileExt);
            // Mark img as 'isparsed'
            img.setAttribute('data-lzld-isparsed','true');
            // Return parsed parts
            return {
                filePath: filePath,
                fileName: fileName,
                fileExt: fileExt,
                aliasBase: filePath + fileName,
                baseSrc: filePath + fileName + fileExt
            };
        },
        getAliasArray: function(img, imgConfig) {
            var that = this;
            var aliasArray = [];
            var useHighRes = that._f.isHighRes && !imgConfig.lowres;
            var viewportWidth = that._u.getViewport("Width");
            var viewportWidthFactor = Math.floor( viewportWidth / imgConfig.widthfactor );
            var firstAlias = !imgConfig.sizedown ? viewportWidthFactor * imgConfig.widthfactor : (viewportWidthFactor+1) * imgConfig.widthfactor;
            // look for at least one resized image
            if (useHighRes) { aliasArray.push( '_' + firstAlias + imgConfig.highressuffix ); }
            aliasArray.push( '_' + firstAlias );
            // if longfallback then add to many more to the array
            if (imgConfig.longfallback) {
                for (var i = viewportWidthFactor-1; i >= 0; i--){
                    var alias = !imgConfig.sizedown ? i * imgConfig.widthfactor : (i+1) * imgConfig.widthfactor;
                    if (alias > 0) {
                        if (useHighRes) { aliasArray.push( '_' + alias + imgConfig.highressuffix ); }
                        aliasArray.push( '_' + alias );
                    }
                }
            }
            // a fallback to look for the smallest image
            if (useHighRes) { aliasArray.push( '_small' + imgConfig.highressuffix ); }
            aliasArray.push( '_small' );
            // final fallback to the original image
            if (useHighRes) { aliasArray.push( imgConfig.highressuffix ); }
            aliasArray.push( '' );
            
            return aliasArray;
        },
        requestImg: function(imgData) {
            var that = this;
            // if not in requestingBaseImgSrc
            if (imgData.aliasArray.length && that._u.inArray(imgData._p.baseSrc, that.requestingBaseImgSrc) === -1) {
                // build image src with first alias
                var imgSrc = imgData._p.aliasBase + imgData.aliasArray[0] + imgData._p.fileExt;
                // remove first alias (the one we just used) from aliasArray so the next time requestImg is called, it will use the next one
                imgData.aliasArray.splice(0,1);
                // store base img src to prevent another request of the same image
                that.requestingBaseImgSrc.push(imgData._p.baseSrc);
                // check if in available before making a request
                if (that._u.inArray(imgSrc, that.availableImgSrc) !== -1) {
                    // remove base img src from requestingBaseImgSrc so it can get in the next time
                    that.requestingBaseImgSrc.splice(that.requestingBaseImgSrc.length-1, 1);
                    // change image
                    that.requestSuccess(imgSrc, imgData);
                    return;
                // check if in unavailable before making request
                } else if (that._u.inArray(imgSrc, that.unavailableImgSrc) !== -1) {
                    // remove base img src from requestingBaseImgSrc so it can get in the next time
                    that.requestingBaseImgSrc.splice(that.requestingBaseImgSrc.length-1, 1);
                    // call requestImg again, but this time the first item of aliasArray will be different
                    that.requestImg(imgData);
                } else { 
                    // make the next request in the alias array
                    that.sendImgRequest(imgSrc, imgData);
                }
            }
        },
        sendImgRequest: function(imgSrc, imgData) {
            var that = this;
            var i = that.xmlhttp.length;
            that.xmlhttp[i] = window.XMLHttpRequest ? new XMLHttpRequest() : new ActiveXObject("Microsoft.XMLHTTP");
            that.xmlhttp[i].onreadystatechange = function() {
                if (that.xmlhttp[i].readyState === 4) {
                    // when it's finished checking it will remove the baseSrc from requestingBaseImgSrc
                    that.requestingBaseImgSrc.splice( that.requestingBaseImgSrc.indexOf( imgData._p.baseSrc ), 1);
                    // Check the status
                    if (that.xmlhttp[i].status === 200) {
                        // add to available imgSrcs
                        that.availableImgSrc.push(imgSrc);
                        // call callback
                        that.requestSuccess(imgSrc, imgData);
                    } else {
                        // add to unavailable imgSrcs
                        that.unavailableImgSrc.push(imgSrc);
                        // try the next alias
                        that.requestImg(imgData);
                    }
                }
            };
            that.xmlhttp[i].open("GET",imgSrc,false);
            that.xmlhttp[i].send();
        },
        requestSuccess: function(imgSrc, imgData) {
            var that = this;
            var img = imgData._e.img;
            // Change the img src to the successful imgSrc
            img.src = imgSrc;
            // add image to list of loadedImgs
            if (that._u.inArray(imgData, that.loadedImgs) === -1) {
                that.loadedImgs.push(imgData);
            }
            // mark as complete
            img.setAttribute('data-lzld-complete','true');
        }
    };

    // Utility methods - methods that are ok all by themselves and don't need any extra outside info
    LazyloadResponsive._u = {
        lzld: LazyloadResponsive,
        _o: config,
        throttle: function(fn, minDelay) {
            var lastCall = 0;
            return function() {
                var now = +new Date();
                if (now - lastCall < minDelay) {
                    return;
                }
                lastCall = now;
                // we do not return anything as
                // https://github.com/documentcloud/underscore/issues/387
                fn.apply(this, arguments);
            };
        },
        // @see: http://ejohn.org/projects/flexible-javascript-events/
        addEvent: function( obj, type, fn ) {
            if ( obj.attachEvent ) {
                obj['e'+type+fn] = fn;
                obj[type+fn] = function(){obj['e'+type+fn]( window.event );};
                obj.attachEvent( 'on'+type, obj[type+fn] );
            } else {
                obj.addEventListener( type, fn, false );
            }
        },
        // https://github.com/jquery/jquery/blob/f3515b735e4ee00bb686922b2e1565934da845f8/src/core.js#L610
        // We cannot use Array.prototype.indexOf because it's not always available
        inArray: function(elem, array, i) {
            var len;
            if ( array ) {
                if ( Array.prototype.indexOf ) {
                    return Array.prototype.indexOf.call( array, elem, i );
                }
                len = array.length;
                i = i ? i < 0 ? Math.max( 0, len + i ) : i : 0;
                for ( ; i < len; i++ ) {
                    // Skip accessing in sparse arrays
                    if ( i in array && array[ i ] === elem ) {
                        return i;
                    }
                }
            }
            return -1;
        },
        isVisible: function(img) {
            var that = this;
            var winH = that.getViewport("Height");
            return (that.contains(document.documentElement, img) && img.getBoundingClientRect().top < winH + that._o.offset);
        },
        getViewport: function(dimension) {
            if (document.documentElement["client"+dimension] >= 0) {
              return document.documentElement["client"+dimension];
            } else if (document.body && document.body["client"+dimension] >= 0) {
              return document.body["client"+dimension];
            } else if (window["inner"+dimension] >= 0) {
              return window["inner"+dimension];
            } else {
              return 0;
            }
        },
        // https://github.com/jquery/sizzle/blob/3136f48b90e3edc84cbaaa6f6f7734ef03775a07/sizzle.js#L708
        contains: function(a,b) {
            if (document.documentElement.compareDocumentPosition) {
                return !!(a.compareDocumentPosition( b ) & 16);
            }
            if (document.documentElement.contains) {
                return a !== b && ( a.contains ? a.contains( b ) : false );
            }
            while ( (b = b.parentNode) ) {
              if ( b === a ) {
                return true;
              }
            }
            return false;
        },
        getPixelRatio: function() {
            return !!window.devicePixelRatio ? window.devicePixelRatio : 1;
        }
    };
    
    // Bind to window
    window.LazyloadResponsive = LazyloadResponsive;
    // Engage!
    if (domReady) {
        window.LazyloadResponsive.initialize();
    } else {
        contentLoaded(window, function(){
            window.LazyloadResponsive.initialize();
        });
    }
    
})(this,document);

src/lazyload-responsive.js
/*
 * lazyload-responsive
 * https://github.com/kevinkirchner/lazyload-responsive
 *
 * Copyright (c) 2013 Kevin Kirchner
 * Licensed under the MIT license.
 */

!(function($, window, undefined) {
    
    "use strict"
    
    /* LAZYLOADRESPONSIVE PUBLIC CLASS DEFINITION
     * ========================================== */
    
    var LazyloadResponsive = function( element, options ) {
        this.init(element, options);
    };
    
    LazyloadResponsive.prototype = {
        constructor: LazyloadResponsive,
        init: function( element, options ) {
            this.$element = $(element);
            this.options = this.getOptions(options);
        },
        getOptions: function( options ) {
            return $.extend({}, $.fn.lazyloadResponsive.defaults, options, this.$element.data());
        }
    };
    
    /* LAZYLOADRESPONSIVE PLUGIN DEFINITION 
     * ==================================== */
    
    // Collection method
    $.fn.lazyloadResponsive = function( option ) {
        return this.each(function () {
            var $this = $(this)
              , data = $this.data('lzldrspv')
              , options = typeof option == 'object' && option
            if (!data) $this.data('lzldrspv', (data = new LazyloadResponsive(this, options)));
            if (typeof option == 'string') data[option]();
        });
    };

    $.fn.lazyloadResponsive.Constructor = LazyloadResponsive;

    // Static method default options.
    $.fn.lazyloadResponsive.defaults = {
        srcAttr: 'data-lzld-src',
        offset: 200, // distance (px) below the fold where images will be loaded
        highResThreshold: 1.5, // any pixel ratio >= this number will be flagged as high-res
        throttleInterval: 20, // throttled interval for scroll and resize events
        useGlobalImgConfig: false, // if `false`, the script will look for the following imgConfig on each lzld img (e.g. img.getAttribute('data-lzld-highressuffix') || imgConfig.highressuffix - which takes the script longer to process); setting to `true` is the fastest option
        findSmallerImgs: true, // forces script to look for and load smaller images as viewport is resized to a smaller size
        imgConfig: {
            highressuffix: '@2x', // e.g. imagename@2x.jpg or imagename_400@2x.jpg would be the high-res images
            loadevent: ['scroll','resize'], // you may want to load the images on a custom event; this is where you do that
            longfallback: true, // will look for all smaller images before loading the original (and largest image)
            lowres: false, // forces script to **not** look for high-res images
            sizedown: false, // by default images are sized up; this option forces it to get an image larger than the viewport and shrink it; NOTE: setting to `true` will load larger images and increase pageload
            widthfactor: 200 // looks for images with the following naming convention [real-image-src]_[factor of widthfactor].[file-extenstion]
        }
    };
    
    /* LAZYLOADRESPONSIVE DATA API
     * =========================== */
    
    $(function () {
        // Onload
        // Scroll
        // Resize
    /*
        $('body').on(function (e) {
            var $this = $(this);
            if ($this.data('lzldrspv')) return;
            e.preventDefault();
            $this.lazyloadResponsive($this.data());
        })
    */
    });

}(jQuery, this));
